src/hooks/useDataSource/useDataSource.tsx

```diff
@@ -1,49 +1,41 @@
 import * as React from "react";
 
-type Refreshable = { refresh: () => void };
-
 type DataSuccess<T> = {
   state: "success";
   data: T;
-} & Refreshable;
+};
 
 type DataError = {
   state: "error";
-} & Refreshable;
+};
 
 type DataLoading = {
   state: "loading";
-} & Refreshable;
+};
+
+type Refreshable = {
+  refresh: () => void;
+};
 
-export type DataResult<T> = DataSuccess<T> | DataError | DataLoading;
+type Data<T> = DataSuccess<T> | DataError | DataLoading;
+export type DataResult<T> = Data<T> & Refreshable;
 
 export function useDataSource<T>(dataSource: () => Promise<T>): DataResult<T> {
-  const [data, setData] = React.useState<T | undefined>();
-  const [loading, setLoading] = React.useState(true);
-  const [error, setError] = React.useState(false);
+  const [data, setData] = React.useState<Data<T>>({
+    state: "loading",
+  });
+
   const refresh = React.useCallback(() => {
-    setLoading(true);
-    setError(false);
+    setData({
+      state: "loading",
+    });
+
     dataSource()
-      .then((newData) => setData(newData))
-      .catch(() => setError(true)) // TODO do something with the caught error?
-      .finally(() => setLoading(false));
+      .then((newData) => setData({ state: "success", data: newData }))
+      .catch(() => setData({ state: "error" }));
   }, []);
 
   React.useEffect(refresh, []);
 
-  if (error) {
-    return { state: "error", refresh };
-  }
-
-  if (loading) {
-    return { state: "loading", refresh };
-  }
-
-  if (data) {
-    return { state: "success", data, refresh };
-  }
-
-  // TODO refactor the above so that this line is not necessary
-  throw new Error("This shouldn't be possible");
+  return data;
```

Comment:
This also needs to return the refresh callback:
```suggestion
  return { ...data, refresh };
```